xdsl/frontend/jaxpr/__init__.py

```diff
@@ -26,58 +47,105 @@ class IRGen:
     """A "module" matches a jaxpr source file: containing a list of functions."""
 
     builder: Builder
+    primitive_builder_jax: JaxPrimitiveBuilder
 
     """
     The builder is a helper class to create IR inside a function. The builder
     is stateful, in particular it keeps an "insertion point": this is where
     the next operations will be introduced."""
 
+    symbol_table: ScopedSymbolTable | None = None
+    """
+    The symbol table maps a variable name to a value in the current scope.
+    Entering a function creates a new scope, and the function arguments are
+    added to the mapping. When the processing of a function is terminated, the
+    scope is destroyed and the mappings created in this scope are dropped."""
+
     def __init__(self):
         # We create an empty MLIR module and codegen functions one at a time and
         # add them to the module.
         self.module = ModuleOp([])
+        self.primitive_builder_jax = JaxPrimitiveBuilder()
         self.builder = Builder.at_end(self.module.body.blocks[0])
 
+    def declare(self, var: str, value: SSAValue) -> bool:
+        """
+        Declare a variable in the current scope, return success if the variable
+        wasn't declared yet."""
+        assert self.symbol_table is not None
+        if var in self.symbol_table:
+            return False
+        self.symbol_table[var] = value
+        return True
+
+    def ir_gen_variable_expr(self, name: str) -> SSAValue:  # pyright: ignore[reportReturnType]
+        """
+        This is a reference to a variable in an expression. The variable is
+        expected to have been declared and so should have a value in the symbol
+        table, otherwise emit an error and return nullptr."""
+        assert self.symbol_table is not None
+        try:
+            variable = self.symbol_table[name]
+            return variable
+        except Exception as e:
+            self.error(f"error: unknown variable `{name}`", e)
```

Comment:
this is also not great, we're just silently ignoring errors, let's propagate them with a new error instead, or just let the compiler crash with the original error